### DISCUSSION TITLE
Use lettuce's native tracing functionality

### DISCUSSION THREAD

**Discussion by SimoneGiusso** [open-telemetry/opentelemetry-java-instrumentation#6345] [Q&A] [answered]
Hi!

I would like to use the lettuce's native tracing functionality to see spans for each redis call and include relevant info in the trace (redis key, etc...). 

I saw that openTelemetry provides already an implementation of the lettuce [Tracing](https://github.com/lettuce-io/lettuce-core/wiki/Tracing) interface [here](https://github.com/open-telemetry/opentelemetry-java-instrumentation/blob/main/instrumentation/lettuce/lettuce-5.1/library/src/main/java/io/opentelemetry/instrumentation/lettuce/v5_1/OpenTelemetryTracing.java). 

However I don't know how to use it since I cannot use this `OpenTelemetryTracing` class and configure tracing through `ClientResources` as suggest in the lettuce documentation since `OpenTelemetryTracing` is not public. 

I have to add some configuration I guess. Can someone help me figure out what I missed?. Btw this is the [demo](https://github.com/SimoneGiusso/Question-OpenTelemetryRedis) I'm using to test this tracing functionality.

**Comment by mateuszrzeszutek** [comment, marked as answer]
Hey @SimoneGiusso ,
The entrypoint of the Lettuce instrumentation library is the [`LettuceTelemetry`](https://github.com/open-telemetry/opentelemetry-java-instrumentation/blob/main/instrumentation/lettuce/lettuce-5.1/library/src/main/java/io/opentelemetry/instrumentation/lettuce/v5_1/LettuceTelemetry.java) class. Once you construct it, you can use the [`newTracing()`](https://github.com/open-telemetry/opentelemetry-java-instrumentation/blob/main/instrumentation/lettuce/lettuce-5.1/library/src/main/java/io/opentelemetry/instrumentation/lettuce/v5_1/LettuceTelemetry.java#L49) method to obtain an instance of Lettuce `Tracing` that you can use on `ClientResource`. See [one of our tests](https://github.com/open-telemetry/opentelemetry-java-instrumentation/blob/main/instrumentation/lettuce/lettuce-5.1/library/src/test/groovy/io/opentelemetry/instrumentation/lettuce/v5_1/LettuceSyncClientTest.groovy#L16) for a brief example.

**Reply by SimoneGiusso** [reply]
Thank you! 

I share how I configured the LettuceConnectionFactory (basic settings):

```java
    @Bean
    public ReactiveRedisConnectionFactory reactiveRedisConnectionFactory(OpenTelemetry openTelemetry) {
        RedisConfiguration redisConfiguration = new RedisStandaloneConfiguration("localhost", 6379);
        LettuceClientConfiguration lettuceClientConfiguration = LettuceClientConfiguration.builder().
            clientResources(ClientResources.builder()
                .tracing(LettuceTelemetry.create(openTelemetry).newTracing())
                .build())
            .build();

        return new LettuceConnectionFactory(redisConfiguration, lettuceClientConfiguration);
    }
```

Where `OpenTelemetry` it's a spring bean autogenerated with `spring-cloud-sleuth-otel-autoconfigure`.

It works!
